doc/wol.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "WOL 1"
.TH WOL 1 2024-03-17 "wol 0.7.1" wol
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
wol \- Wake On LAN client
.SH SYNOPSIS
.IX Header "SYNOPSIS"
wol [\fIOPTION\fR] ... \fIMAC-ADDRESS\fR ...
.SH DESCRIPTION
.IX Header "DESCRIPTION"
This manual gives you an introduction into \f(CW\*(C`wol\*(C'\fR, the Wake On LAN client.
It remotely turns on computers that supports Magic Packet technology
(also known as Wake On LAN).
.PP
Some workstations support SecureON password feature. These machines only
wake up if you provide the correct password. \f(CW\*(C`wol\*(C'\fR also provides this
feature.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-\-help\fR 4
.IX Item "--help"
Print a summary of the command line options.
.IP \fB\-V\fR 4
.IX Item "-V"
.PD 0
.IP \fB\-\-version\fR 4
.IX Item "--version"
.PD
Print the version number of \f(CW\*(C`wol\*(C'\fR.
.IP "\fB\-h HOST\fR" 4
.IX Item "-h HOST"
.PD 0
.IP \fB\-\-host=HOST\fR 4
.IX Item "--host=HOST"
.IP "\fB\-i HOST\fR" 4
.IX Item "-i HOST"
.IP \fB\-\-ipaddr=HOST\fR 4
.IX Item "--ipaddr=HOST"
.PD
Broadcast packet to this IP address or hostname. This is important if your wol
client is a multihomed host and you want to send only to one subnet
(default IP address is \f(CW255.255.255.255\fR).
.IP "\fB\-p NUM\fR" 4
.IX Item "-p NUM"
.PD 0
.IP \fB\-\-port=NUM\fR 4
.IX Item "--port=NUM"
.PD
Send packet with this destination port \fINUM\fR. This option is important if
your packet filter would block the default destination port \f(CW40000\fR.
.IP "\fB\-f FILE\fR" 4
.IX Item "-f FILE"
.PD 0
.IP \fB\-\-file=FILE\fR 4
.IX Item "--file=FILE"
.PD
Read hardware addresses, IP addresses/hostnames, optional ports and SecureON
password from file \fIFILE\fR. If \fIFILE\fR is \fI\-\fR \f(CW\*(C`wol\*(C'\fR reads from
stdin.
.IP \fB\-\fR 4
.IX Item "-"
Read hardware addresses, IP addresses, optional ports and optional SecureON
password from stdin. So you can extract \fIMAC-ADDRESSES\fR from any source
and pipe the data into \f(CW\*(C`wol\*(C'\fR so.
.IP \fB\-v\fR 4
.IX Item "-v"
.PD 0
.IP \fB\-\-verbose\fR 4
.IX Item "--verbose"
.PD
Turns on verbose output.
.IP "\fB\-w NUM\fR" 4
.IX Item "-w NUM"
.PD 0
.IP \fB\-\-wait=NUM\fR 4
.IX Item "--wait=NUM"
.PD
Waits \fINUM\fR milliseconds between Magic Packets. Also known as fuse health
pack.
.IP \fB\-\-passwd[=PASS]\fR 4
.IX Item "--passwd[=PASS]"
Send a magic packet with SecureON password feature. \fIPASS\fR is written as
x\-x-x-x-x-x, where x is a hexadecimal number between 0 and ff which represents
one byte of the password.
.Sp
If you don't provide \fIPASS\fR \f(CW\*(C`wol\*(C'\fR prompts you for a password.
.Sp
To set the password of your SecureON capable NIC, you can use
\&\f(CW\*(C`ethtool\*(C'\fR (<\fBhttp://www.sourceforge.net/projects/gkernel/\fR>).
.SH BUGS
.IX Header "BUGS"
Please report bugs to <\fBkrennwallner@aon.at\fR>. Feel free to send bug
reports, translations, improvement suggestions and patches to this email
address.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
GNU info entry for \fIwol\fR.
.SH AUTHOR
.IX Header "AUTHOR"
Written by Thomas Krennwallner <krennwallner@aon.at>.
.SH COPYRIGHT
.IX Header "COPYRIGHT"
Copyright (C) 2001,2002,2003,2004 Thomas Krennwallner <\fBkrennwallner@aon.at\fR>
.PP
This document may be distributed and modified under the terms of the GNU
General Public License.
.PP
This manual is free software; you may redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2, or (at your option) any later version.
.PP
This manual is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for a
particular purpose. See the GNU General Public License for more details.
.PP
A copy of the GNU General Public License is available on the World Wide Web at
the GNU website. You can also obtain it by writing to the Free Software
Foundation, Inc., 59 Temple Place \- Suite 330, Boston, MA 02111\-1307, USA.


src/wol.c
/*
 * wol - wake on lan client
 *
 * main program
 *
 * $Id: wol.c,v 1.19 2004/05/08 09:25:45 wol Exp $
 *
 * Copyright (C) 2000,2001,2002,2003,2004 Thomas Krennwallner <krennwallner@aon.at>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 */

#include <errno.h>
#include <error.h>
#include <getopt.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "macfile.h"
#include "magic.h"
#include "net.h"
#include "proxy.h"
#include "wol.h"

/* pointer to a MAC address */
static char *mac_str = NULL;

/* IP Address or hostname magic packet is addressed to */
static char *host_str = DEFAULT_IPADDR;

/* filename with mac addresses */
static char *pathname = NULL;

/* udp port */
static unsigned int port = DEFAULT_PORT;

/* SecureON password */
static char *passwd = NULL;

/* default is not to read from stdin */
static int request_stdin = 0;

/* be verbose */
static int verbose = 0;

/* send udp / raw / proxy packet */

#define UDP_MODE 0x1
#define RAW_MODE 0x2
#define PROXY_MODE 0x4

static unsigned int packet_mode = UDP_MODE;

/* how long to wait between packets */
static unsigned int msecs = 0;

/* a magic packet */
static struct magic *magic = NULL;

/* socket file descriptor */
static int sockfd = -1;

static void usage(int status) {
        if (status) {
                fprintf(stderr, _("Try `%s --help' for more information.\n"), program_invocation_name);
        } else {
                fprintf(stdout,
                        _("\
Usage: %s [OPTION] ... MAC-ADDRESS ...\n\
Wake On LAN client - wakes up magic packet compliant machines.\n\n\
    --help          display this help and exit\n\
-V, --version       output version information and exit\n\
-v, --verbose       verbose output\n\
-w, --wait=NUM      wait NUM millisecs after sending\n\
-h, --host=HOST     broadcast to this IP address or hostname\n\
-i, --ipaddr=HOST   same as --host\n\
-p, --port=NUM      broadcast to this UDP port\n\
-f, --file=FILE     read addresses from file FILE (\"-\" reads from stdin)\n\
    --passwd[=PASS] send SecureON password PASS (if no PASS is given, you\n\
                    will be prompted for the password)\n\
-r, --raw           send raw ethernet magic packet\n\
-s, --proxy=HOST    send wake up information to wolp proxy HOST\n\
-u, --udp           send udp magic packet\n\
\n\
Each MAC-ADDRESS is written as x:x:x:x:x:x, where x is a hexadecimal number\n\
between 0 and ff which represents one byte of the address, which is in\n\
network byte order (big endian).\n"),
                        program_invocation_name);

                fprintf(stdout, _("\n\
PASS is written as x-x-x-x-x-x, where x is a hexadecimal number between 0\n\
and ff which represents one byte of the password.\n"));

                fprintf(stdout, _("\nReport bugs to <krennwallner@aon.at>\n"));
        }

        exit(status);
}

static void version(void) {
        fprintf(stdout, PACKAGE " " VERSION "\n\n");
        fprintf(stdout, _("\
Copyright (C) 2000-2004 Thomas Krennwallner <krennwallner@aon.at>\n\
This is free software; see the source for copying conditions. There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\
\n"));
}

/* parse command line and set various globals */
static int parse_args(int argc, char *argv[]) {
        int c;
        int option_index;
        int password_set = 0;
        char *options = "Vvw:h:i:p:f:s:ru-";
        static struct option long_options[] = {
                {   "help",       no_argument, NULL, 'H'},
                {"version",       no_argument, NULL, 'V'},
                {"verbose",       no_argument, NULL, 'v'},
                {   "wait", required_argument, NULL, 'w'},
                {   "host", required_argument, NULL, 'h'},
                { "ipaddr", required_argument, NULL, 'i'},
                {   "port", required_argument, NULL, 'p'},
                {   "file", required_argument, NULL, 'f'},
                { "passwd", optional_argument, NULL, 'P'},
                {  "proxy", required_argument, NULL, 's'},
                {    "raw",       no_argument, NULL, 'r'},
                {    "udp",       no_argument, NULL, 'u'},
                {     NULL,                 0, NULL,   0}
        };

        if (argc == 1) {
                error(0, 0, _("Too few arguments."));
                usage(1);
        }

        for (;;) {
                c = getopt_long(argc, argv, options, long_options, &option_index);
                if (c == -1)
                        break;

                switch (c) {
                case 'H':
                        usage(0);
                        break;

                case 'V':
                        version();
                        exit(0);
                        break;

                case 'v':
                        verbose = 1;
                        break;

                case 'w':
                        if (sscanf(optarg, "%u", &msecs) != 1) {
                                error(0, 0, _("Invalid time given"));
                                usage(1);
                        }
                        msecs *= 1000;
                        break;

                case 'u':
                        packet_mode |= UDP_MODE;
                        packet_mode &= ~(RAW_MODE | PROXY_MODE);
                        break;

                case 'r':
                        packet_mode |= RAW_MODE;
                        packet_mode &= ~(UDP_MODE | PROXY_MODE);
                        break;

                case 's':
                        packet_mode |= PROXY_MODE;
                        packet_mode &= ~(UDP_MODE | RAW_MODE);
                        break;
                case 'h':
                case 'i':
                        host_str = optarg;
                        break;

                case 'p':
                        if ((sscanf(optarg, "%5u", &port) != 1) || port > 65535 || port == 0) {
                                error(0, 0, _("Invalid port given"));
                                usage(1);
                        }
                        break;

                case 'f':
                        pathname = optarg;
                        break;

                case 'P':
                        if (optarg == NULL) {
                                if (password_set)
                                        break;

                                if ((passwd = getpass(_("Password"))) == NULL) {
                                        error(1, 0, "getpass failed");
                                }
                                password_set = 1;
                        } else {
                                passwd = optarg;
                        }
                        break;

                case '?':
                        break;
                }
        }

        if ((optind == argc) && (pathname == NULL)) {
                error(0, 0, _("You must specify at least one MAC-ADDRESS."));
                usage(1);
        }

        /* check if stdin is requested */
        if (optind < argc) {
                int i;

                for (i = optind; i < argc; ++i) {
                        if (argv[i][0] == '-' && argv[i][1] == 0) {
                                request_stdin = 1;
                                break;
                        }
                }
        } else if (pathname != NULL) {
                if (!strncmp(pathname, "-", 1)) {
                        request_stdin = 1;
                }
        }

        /* return the offset of the GNU getopt sorted parameters */
        return optind;
}

static int assemble_and_send(
                struct magic *m,
                const char *mac_str,
                const char *host_str,
                unsigned int portnum,
                const char *pass_str,
                int socketfd) {
        if (packet_mode & UDP_MODE) {
                int ret = magic_assemble(m, mac_str, pass_str);

                switch (ret) {
                case -1:
                        error(0, errno, _("Cannot assemble magic packet for '%s'"), mac_str);
                        errno = 0;
                        return -1;

                case -2:
                        error(0, 0, _("Invalid password given for '%s'"), mac_str);
                        errno = 0;
                        return -1;
                }

                if (udp_send(socketfd, host_str, portnum, m->packet, m->size)) {
                        error(0, errno, _("Cannot send magic packet for '%s' to %s:%d"), mac_str, host_str, portnum);
                        errno = 0;
                        return -1;
                }
        } else if (packet_mode & PROXY_MODE) {
                char *proxy_pass = (char *) pass_str;
                int errval;

                /* request proxy password if it's not given */
                if (proxy_pass == NULL) {
                        if ((proxy_pass = getpass(_("Password"))) == NULL) {
                                error(1, 0, "getpass4 failed");
                        }
                }

                errval = proxy_send(socketfd, mac_str, proxy_pass);

                /* free possible new password */
                free(proxy_pass);
                proxy_pass = NULL;

                if (errval) {
                        error(0, errno, _("Cannot send magic packet for '%s' to %s:%d"), mac_str, host_str, portnum);
                        errno = 0;
                        return -1;
                }
        } else /* RAW_MODE */
        {
                int ret = magic_assemble(m, mac_str, pass_str);

                switch (ret) {
                case -1:
                        error(0, errno, _("Cannot assemble magic packet for '%s'"), mac_str);
                        errno = 0;
                        return -1;

                case -2:
                        error(0, 0, _("Invalid password given for '%s'"), mac_str);
                        errno = 0;
                        return -1;
                }

                raw_send(socketfd, m->packet, m->size);
        }

        fprintf(stdout, _("Waking up %s"), mac_str);
        if (verbose) {
                fprintf(stdout, _(" with %s:%d"), host_str, portnum);
        }
        fprintf(stdout, _("...\n"));

        if (msecs) {
                usleep(msecs);
        }

        return 0;
}

int main(int argc, char *argv[]) {
        int i;
        int ret = 0;

        setlocale(LC_ALL, "");
        bindtextdomain(PACKAGE, LOCALEDIR);
        textdomain(PACKAGE);

        i = parse_args(argc, argv);

        magic = magic_create(passwd != NULL);
        if (magic == NULL) {
                exit(1);
        }

        if (packet_mode & UDP_MODE) {
                sockfd = udp_open();
        } else if (packet_mode & PROXY_MODE) {
                sockfd = tcp_open(host_str, port);
        } else /* RAW_MODE */
        {
                sockfd = raw_open();
        }

        if (sockfd < 0) {
                exit(1);
        }

        /* loop through possible MAC addresses */
        if (!request_stdin) {
                for (; i < argc; i++) {
                        ret -= assemble_and_send(magic, argv[i], host_str, port, passwd, sockfd);
                }
        }

        /* -f given */
        if (pathname || request_stdin) {
                FILE *fp;

                if (request_stdin) {
                        fp = stdin;
                } else {
                        fp = fopen(pathname, "re");
                        if (fp == NULL) {
                                error(1, errno, "%s", pathname);
                        }
                }

                /* loop through fp */
                for (;;) {
                        if (macfile_parse(fp, &mac_str, &host_str, &port, &passwd))
                                break;

                        if (port == 0 || port > 65535) {
                                port = DEFAULT_PORT;
                        }

                        ret -= assemble_and_send(magic, mac_str, host_str, port, passwd, sockfd);

                        free(mac_str);
                        free(host_str);
                        free(passwd);
                }

                fclose(fp);
        }

        close(sockfd);

        magic_destroy(magic);

        exit(ret != 0);
}
